### PR TITLE
DDS default_datareader_qos

### PR DESCRIPTION
**Pull request by IkerLuengo**
- Change ReaderQos for DataReaderQos
- DataReaderQos methods on PIM
- unit test for get/set_default_datareader_qos on PIM
- link PSM datareader_qos getter/setter to PIM implementation
- unit test for PSM datareader_qos getter/setter

NOTE: This PR must be merged after #1122

### CODE REVIEW

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 1]
```diff
@@ -627,6 +627,189 @@ TypeSupport DataReaderImpl::type()
...
+    }
+    if (first_time && !(qos.resource_limits() == new_qos.resource_limits()))
+    {
+        qos.resource_limits() = new_qos.resource_limits();
+        qos.resource_limits().hasChanged = true;
+    }
+    if (!(qos.reader_data_lifecycle() == new_qos.reader_data_lifecycle()))
+    {
+        qos.reader_data_lifecycle() = new_qos.reader_data_lifecycle();
+        qos.reader_data_lifecycle().hasChanged = true;
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp]:
```suggestion
```
ReaderDataLifecycleQos doesn't need to inherit from QosPolicy as it don't have to be sent

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 2]
```diff
@@ -627,6 +627,189 @@ TypeSupport DataReaderImpl::type()
...
+    }
+
+    if (qos.expectsInlineQos() != new_qos.expectsInlineQos())
+    {
+        qos.expectsInlineQos(new_qos.expectsInlineQos());
+    }
+
+    if (!(qos.properties() == new_qos.properties()))
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp]:
```suggestion
    if (first_time && !(qos.properties() == new_qos.properties()))
```

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 3]
```diff
@@ -627,6 +627,189 @@ TypeSupport DataReaderImpl::type()
...
+    {
+        qos.properties() = new_qos.properties();
+    }
+
+    if (!(qos.enpoint() == new_qos.enpoint()))
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp]:
```suggestion
    if (first_time && !(qos.enpoint() == new_qos.enpoint()))
```

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 4]
```diff
@@ -627,6 +627,189 @@ TypeSupport DataReaderImpl::type()
...
+    {
+        qos.enpoint() = new_qos.enpoint();
+        qos.enpoint().hasChanged = true;
+    }
+
+    if (!(qos.reader_resource_limits() == new_qos.reader_resource_limits()))
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp]:
```suggestion
    if (first_time && !(qos.reader_resource_limits() == new_qos.reader_resource_limits()))
```

File: include/fastdds/dds/core/policy/ReaderDataLifecycleQosPolicy.hpp [hunk 5]
```diff
@@ -14,37 +14,51 @@
...
+
 using eprosima::fastrtps::Duration_t;
 
 namespace eprosima {
 namespace fastdds {
 namespace dds {
 
 //! @brief A struct storing the base status
-class ReaderDataLifecycleQosPolicy
+class ReaderDataLifecycleQosPolicy : public QosPolicy
```
Comment (laura-eprosima) [inline include/fastdds/dds/core/policy/ReaderDataLifecycleQosPolicy.hpp:33]:
```suggestion
class ReaderDataLifecycleQosPolicy
```

File: include/fastdds/dds/subscriber/qos/DataReaderQos.hpp [hunk 6]
```diff
@@ -20,13 +20,168 @@
...
+#include <fastrtps/attributes/TopicAttributes.h>
+
+
 namespace eprosima {
 namespace fastdds {
 namespace dds {
 
+using TopicAttributesQos = fastrtps::TopicAttributes;
+
+class RTPSReliableReaderQos : public QosPolicy
```
Comment (laura-eprosima) [inline include/fastdds/dds/subscriber/qos/DataReaderQos.hpp:38]:
```suggestion
class RTPSReliableReaderQos
```

File: include/fastdds/dds/subscriber/qos/DataReaderQos.hpp [hunk 7]
```diff
@@ -20,13 +20,168 @@
...
+    fastrtps::rtps::LocatorList_t remoteLocatorList;
+
+    //!Underlying History memory policy (Extension).
+    fastrtps::rtps::MemoryManagementPolicy_t historyMemoryPolicy;
+
+    //!User Defined ID, used for StaticEndpointDiscovery (Extension).
+    int16_t m_userDefinedID;
+
+    //!Entity ID, to specify the EntityID of the enpoint (Extension).
+    int16_t m_entityID;
```
Comment (laura-eprosima) [inline include/fastdds/dds/subscriber/qos/DataReaderQos.hpp:122]: Coding style

File: include/fastdds/dds/subscriber/qos/DataReaderQos.hpp [hunk 8]
```diff
@@ -20,13 +20,168 @@
...
+
+    /*!
+     * @brief Times associated with the Reliable Readers events.
+     */
+    fastrtps::rtps::ReaderTimes reader_times;
+
+    /*!
+     * @brief Control the sending of positive ACKs
+     */
+    DisablePositiveACKsQosPolicy disablePositiveACKs;
```
Comment (laura-eprosima) [inline include/fastdds/dds/subscriber/qos/DataReaderQos.hpp:71]: Coding style

File: include/fastdds/dds/subscriber/qos/DataReaderQos.hpp [hunk 9]
```diff
@@ -20,13 +20,168 @@
...
+    fastrtps::rtps::MemoryManagementPolicy_t historyMemoryPolicy;
+
+    //!User Defined ID, used for StaticEndpointDiscovery (Extension).
+    int16_t m_userDefinedID;
+
+    //!Entity ID, to specify the EntityID of the enpoint (Extension).
+    int16_t m_entityID;
+};
+
+class ReaderResourceLimitsQos : public QosPolicy
```
Comment (laura-eprosima) [inline include/fastdds/dds/subscriber/qos/DataReaderQos.hpp:125]:
```suggestion
class ReaderResourceLimitsQos
```

File: include/fastdds/dds/subscriber/qos/DataReaderQos.hpp [hunk 10]
```diff
@@ -37,84 +192,690 @@ namespace dds {
...
+
+    //!Lifespan Qos (Extension).
+    LifespanQosPolicy lifespan_;
+
+    //!Topic Data Qos (Extension).
+    TopicDataQosPolicy topicData_;
+
+    //!Durability Service Qos (Extension).
+    DurabilityServiceQosPolicy durabilityService_;
+
```
Comment (laura-eprosima) [inline include/fastdds/dds/subscriber/qos/DataReaderQos.hpp:853]: Coding style
```suggestion
    TopicDataQosPolicy topic_data_;

    //!Durability Service Qos (Extension).
    DurabilityServiceQosPolicy durability_service_;

```

File: include/fastdds/dds/subscriber/qos/DataReaderQos.hpp [hunk 11]
```diff
@@ -37,84 +192,690 @@ namespace dds {
...
+    //! Tipe consistency (Extension)
+    TypeConsistencyQos type_consistency_;
+
+    //from SubscriberAttributes
+
+    //!Topic Attributes (Extension).
+    //TopicAttributesQos topic;
+
+    //!Expects Inline QOS (Extension).
+    bool expectsInlineQos_;
```
Comment (laura-eprosima) [inline include/fastdds/dds/subscriber/qos/DataReaderQos.hpp:866]: Coding style

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 12]
```diff
@@ -627,6 +627,189 @@ TypeSupport DataReaderImpl::type()
...
+
+    if (!(to.properties() == from.properties()))
+    {
+        to.properties() = from.properties();
+    }
+
+    if (!(to.enpoint() == from.enpoint()))
+    {
+        to.enpoint() = from.enpoint();
+        to.enpoint().hasChanged = true;
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:802]:
```suggestion
```

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 13]
```diff
@@ -627,6 +627,189 @@ TypeSupport DataReaderImpl::type()
...
+    }
+
+    if (!(to.reader_resource_limits() == from.reader_resource_limits()))
+    {
+        to.reader_resource_limits() = from.reader_resource_limits();
+        to.reader_resource_limits().hasChanged = true;
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:808]:
```suggestion
```

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 14]
```diff
@@ -627,6 +627,189 @@ TypeSupport DataReaderImpl::type()
...
+    }
+    if (first_time && !(to.resource_limits() == from.resource_limits()))
+    {
+        to.resource_limits() = from.resource_limits();
+        to.resource_limits().hasChanged = true;
+    }
+    if (!(to.reader_data_lifecycle() == from.reader_data_lifecycle()))
+    {
+        to.reader_data_lifecycle() = from.reader_data_lifecycle();
+        to.reader_data_lifecycle().hasChanged = true;
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:786]:
```suggestion
```

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 15]
```diff
@@ -627,6 +627,189 @@ TypeSupport DataReaderImpl::type()
...
+    }
+
+    if (to.expectsInlineQos() != from.expectsInlineQos())
+    {
+        to.expectsInlineQos(from.expectsInlineQos());
+    }
+
+    if (!(to.properties() == from.properties()))
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:794]:
```suggestion
    if (first_time && !(to.properties() == from.properties()))
```

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 16]
```diff
@@ -627,6 +627,189 @@ TypeSupport DataReaderImpl::type()
...
+    {
+        to.properties() = from.properties();
+    }
+
+    if (!(to.enpoint() == from.enpoint()))
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:799]:
```suggestion
    if (first_time && !(to.enpoint() == from.enpoint()))
```

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 17]
```diff
@@ -627,6 +627,189 @@ TypeSupport DataReaderImpl::type()
...
+    {
+        to.enpoint() = from.enpoint();
+        to.enpoint().hasChanged = true;
+    }
+
+    if (!(to.reader_resource_limits() == from.reader_resource_limits()))
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:805]:
```suggestion
    if (first_time && !(to.reader_resource_limits() == from.reader_resource_limits()))
```

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 18]
```diff
@@ -627,6 +627,189 @@ TypeSupport DataReaderImpl::type()
...
+    {
+        to.reliability() = from.reliability();
+        to.reliability().hasChanged = true;
+    }
+    if (first_time && to.ownership().kind != from.ownership().kind)
+    {
+        to.ownership() = from.ownership();
+        to.ownership().hasChanged = true;
+    }
+    if (to.destination_order().kind != from.destination_order().kind)
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:732]:
```suggestion
    if (first_time && to.destination_order().kind != from.destination_order().kind)
```

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 19]
```diff
@@ -130,7 +131,7 @@ ReturnCode_t DataReaderImpl::read_next_sample(
 {
     auto max_blocking_time = std::chrono::steady_clock::now() +
 #if HAVE_STRICT_REALTIME
-            std::chrono::microseconds(::TimeConv::Time_t2MicroSecondsInt64(qos_.m_reliability.max_blocking_time));
+            std::chrono::microseconds(::TimeConv::Time_t2MicroSecondsInt64(qos_.reliability.max_blocking_time));
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:134]:
```suggestion
            std::chrono::microseconds(::TimeConv::Time_t2MicroSecondsInt64(qos_.reliability().max_blocking_time));
```

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 20]
```diff
@@ -147,7 +148,7 @@ ReturnCode_t DataReaderImpl::take_next_sample(
...
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:151]:
```suggestion
            std::chrono::microseconds(::TimeConv::Time_t2MicroSecondsInt64(qos_.reliability().max_blocking_time));
```

File: include/fastdds/dds/subscriber/qos/DataReaderQos.hpp (hunk 7 above)
Comment (IkerLuengo) [reply include/fastdds/dds/subscriber/qos/DataReaderQos.hpp:122]: I don't see anything wrong. What specifically?

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1831/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1578/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1825/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1839/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1585/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1833/)

File: include/fastdds/dds/core/policy/QosPolicies.hpp [hunk 21]
```diff
@@ -71,7 +74,6 @@ class QosPolicy
     bool operator ==(
             const QosPolicy& b) const
     {
-        // hastChanged field isn't needed to be compared to being equal two QosPolicy objects.
```
Comment (laura-eprosima) [inline include/fastdds/dds/core/policy/QosPolicies.hpp:74]: Don't remove this comment

File: include/fastdds/dds/core/policy/QosPolicies.hpp [hunk 22]
```diff
@@ -466,8 +467,7 @@ class ReliabilityQosPolicy : public Parameter_t, public QosPolicy
     {
         return (this->kind == b.kind) &&
                (this->max_blocking_time == b.max_blocking_time) &&
-               Parameter_t::operator ==(b) &&
```
Comment (laura-eprosima) [inline include/fastdds/dds/core/policy/QosPolicies.hpp:469]: Why did you remove this line?

File: src/cpp/dds/domain/DomainParticipant.cpp [hunk 23]
```diff
@@ -133,17 +133,25 @@ DomainParticipant& DomainParticipant::default_publisher_qos(
...
+    if ( result == ReturnCode_t::RETCODE_INCONSISTENT_POLICY)
+    {
+        throw dds::core::InconsistentPolicyError("Inconsistent Qos");
+    }
+    if ( result == ReturnCode_t::RETCODE_IMMUTABLE_POLICY)
+    {
+        throw dds::core::ImmutablePolicyError("Inmutable Qos");
+    }
+    return *this;
+}
```
Comment (laura-eprosima) [inline src/cpp/dds/domain/DomainParticipant.cpp:154]: These changes belong to #1122

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 24]
```diff
@@ -627,6 +639,185 @@ TypeSupport DataReaderImpl::type()
...
+    if (to.topicData().getValue() != from.topicData().getValue())
+    {
+        to.topicData() = from.topicData();
+        to.topicData().hasChanged = true;
+    }
+    if (first_time || !(to.durabilityService() == from.durabilityService()))
+    {
+        to.durabilityService() = from.durabilityService();
+        to.durabilityService().hasChanged = true;
+    }
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:768]:
```suggestion
    if (to.topic_data().getValue() != from.topic_data().getValue())
    {
        to.topic_data() = from.topic_data();
        to.topic_data().hasChanged = true;
    }
    if (first_time || !(to.durability_service() == from.durability_service()))
    {
        to.durability_service() = from.durability_service();
        to.durability_service().hasChanged = true;
    }
```

File: src/cpp/fastdds/subscriber/DataReaderImpl.hpp [hunk 25]
```diff
@@ -200,6 +200,31 @@ class DataReaderImpl
     //! Remove all listeners in the hierarchy to allow a quiet destruction
     void disable();
 
+    /* Check whether values in the DataReaderQos are compatible among them or not
+     * @return True if correct.
+     */
+    static ReturnCode_t check_qos (const DataReaderQos& qos);
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/DataReaderImpl.hpp:206]:
```suggestion
    static ReturnCode_t check_qos (
            const DataReaderQos& qos);
```

File: include/fastdds/dds/subscriber/qos/DataReaderQos.hpp [hunk 26]
```diff
@@ -37,84 +137,690 @@ namespace dds {
...
+            const LifespanQosPolicy& new_value)
+    {
+        lifespan_ = new_value;
+    }
+
+    /**
+     * Getter for TopicDataQosPolicy
+     * @return TopicDataQosPolicy reference
+     */
+    TopicDataQosPolicy& topicData()
```
Comment (laura-eprosima) [inline include/fastdds/dds/subscriber/qos/DataReaderQos.hpp:541]:
```suggestion
    TopicDataQosPolicy& topic_data()
```

Comment (laura-eprosima) [reply include/fastdds/dds/subscriber/qos/DataReaderQos.hpp:541]: Same for the other getter and setter

File: include/fastdds/dds/subscriber/qos/DataReaderQos.hpp [hunk 27]
```diff
@@ -37,84 +137,690 @@ namespace dds {
...
+            const TopicDataQosPolicy& new_value)
+    {
+        topic_data_ = new_value;
+    }
+
+    /**
+     * Getter for DurabilityServiceQosPolicy
+     * @return DurabilityServiceQosPolicy reference
+     */
+    DurabilityServiceQosPolicy& durabilityService()
```
Comment (laura-eprosima) [inline include/fastdds/dds/subscriber/qos/DataReaderQos.hpp:569]:
```suggestion
    DurabilityServiceQosPolicy& durability_service()
```
The same for the other getter and setter

File: include/fastdds/dds/core/policy/QosPolicies.hpp (hunk 22 above)
Comment (IkerLuengo) [reply include/fastdds/dds/core/policy/QosPolicies.hpp:469]: This was changed before the `QosPolicy::operator ==()` was corrected, and the I forgot about it. Good catch.

File: src/cpp/dds/domain/DomainParticipant.cpp (hunk 23 above)
Comment (IkerLuengo) [reply src/cpp/dds/domain/DomainParticipant.cpp:154]: I will patch the commit with comment 

> Refs 8113. Implement DataReaderQos methods

to avoid future problems with the merges

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1861/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1603/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1855/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1871/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1612/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1864/)

File: include/fastdds/dds/subscriber/qos/DataReaderQos.hpp [hunk 28]
```diff
@@ -37,84 +137,690 @@ namespace dds {
...
-     * @return True if they can be updated.
-    RTPS_DllAPI bool canQosBeUpdated(
-            const DataReaderQos& qos) const;
-    */
+
+    //!Lifespan Qos (Extension).
+    LifespanQosPolicy lifespan_;
+
+    //!Topic Data Qos (Extension).
+    TopicDataQosPolicy topic_data_;
```
Comment (laura-eprosima) [inline include/fastdds/dds/subscriber/qos/DataReaderQos.hpp:794]:
```suggestion
```

File: include/fastdds/dds/subscriber/qos/DataReaderQos.hpp [hunk 29]
```diff
@@ -37,84 +137,690 @@ namespace dds {
...
+    //!Reliable reader configuration (Extension)
+    RTPSReliableReaderQos reliable_reader_qos_;
+
+    //! Tipe consistency (Extension)
+    TypeConsistencyQos type_consistency_;
+
+    //from SubscriberAttributes
+
+    //!Topic Attributes (Extension).
+    //TopicAttributesQos topic;
```
Comment (laura-eprosima) [inline include/fastdds/dds/subscriber/qos/DataReaderQos.hpp:808]:
```suggestion

```

File: include/fastdds/dds/subscriber/qos/DataReaderQos.hpp (hunk 28 above)
Comment (laura-eprosima) [reply include/fastdds/dds/subscriber/qos/DataReaderQos.hpp:794]: Retrieve this information from TopicQos

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1896/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1634/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1889/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1904/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1642/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1897/)
